src/3rd_party_bundle_add.c

```diff
@@ -94,6 +98,174 @@ static bool parse_options(int argc, char **argv)
+	if (!sys_file_exists(bin_directory)) {
+		if (mkdir_p(bin_directory)) {
+			ret_code = SWUPD_COULDNT_CREATE_DIR;
+			goto close_and_exit;
+		}
+	}
+
+	if (!is_dir(bin_directory)) {
+		error("The path %s for 3rd-party content exists but is not a directory\n", bin_directory);
+		ret_code = SWUPD_COULDNT_CREATE_DIR;
+		goto close_and_exit;
+	}
+
+	fp = fopen(script, "w");
+	if (!fp) {
+		error("The file %s failed to be created\n", script);
+		ret_code = SWUPD_COULDNT_CREATE_FILE;
+		goto close_and_exit;
+	}
+
+	/* get the current values of the environment */
+	env_path = getenv("PATH");
+	env_ld_path = getenv("LD_LIBRARY_PATH");
+
+	/* get the path for the 3rd-party content */
+	third_party_bin_path = str_or_die("%sbin:%susr/bin:%susr/local/bin", globals.path_prefix, globals.path_prefix, globals.path_prefix);
+	third_party_ld_path = str_or_die("%susr/lib64:%susr/local/lib64", globals.path_prefix, globals.path_prefix);
+
+	fprintf(fp, "#!/bin/bash\n");
+	fprintf(fp, "\n");
+	fprintf(fp, "export PATH=%s%s%s\n", third_party_bin_path, env_path ? ":" : "", env_path ? env_path : "");
+	fprintf(fp, "export LD_LIBRARY_PATH=%s%s%s\n", third_party_ld_path, env_ld_path ? ":" : "", env_ld_path ? env_ld_path : "");
+	fprintf(fp, "\n");
+	fprintf(fp, "%s \"$@\"\n", binary);
+
+	/* set the file mode to 0755 */
+	if (chmod(script, mode) < 0) {
+		error("Unable to set permissions for file %s\n", script);
+		ret_code = SWUPD_COULDNT_WRITE_FILE;
+	}
+
+close_and_exit:
+	if (fp) {
+		fclose(fp);
+	}
+	free_string(&binary);
+	free_string(&script);
+	free_string(&bin_directory);
+	free_string(&third_party_bin_path);
+	free_string(&third_party_ld_path);
+
+	return ret_code;
+}
+
+static enum swupd_code validate_binaries(const char *filename)
+{
+	enum swupd_code ret_code = SWUPD_OK;
+	char *bin_directory = NULL;
+	char *script = NULL;
+
+	char *tmp1 = sys_dirname(globals.path_prefix);
+	char *tmp2 = sys_dirname(tmp1);
+	bin_directory = sys_path_join(tmp2, "/bin");
+	free_string(&tmp1);
+	free_string(&tmp2);
+
+	script = sys_path_join(bin_directory, sys_basename(filename));
+
+	/* if a file already exists, report the problem to the user */
+	if (sys_file_exists(script)) {
+		error("There is already a binary called %s in %s\n", sys_basename(filename), bin_directory);
+		ret_code = SWUPD_INVALID_BINARY;
+	}
+	free_string(&script);
+	free_string(&bin_directory);
+
+	return ret_code;
+}
+
+static enum swupd_code process_bundle_binaries(struct list *installed_files, const char *msg, const char *step, proc_binary_fn_t proc_binary_fn)
+{
+	enum swupd_code ret, ret_code = SWUPD_OK;
+	struct list *iter = NULL;
+	struct file *file = NULL;
+	char *filename = NULL;
+	int count = 0;
+	int number_of_files = list_len(installed_files);
+
+	info("%s", msg);
+
+	progress_next_step(step, PROGRESS_BAR);
+	for (iter = list_head(installed_files); iter; iter = iter->next) {
+		file = iter->data;
+		filename = file->filename;
+		if (is_binary(filename) && file->is_exported) {
```

Comment:
Check if is_exported() before checking if is_binary() because it's way faster.

Reply:
good point, actually I had it like that initially but switched the positions so I could comment out the `file->is_exported` condition easier when testing locally. I will change that.